main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -899,6 +875,49 @@ object Lexer {
     TokenKind.Err(LexerError.UnterminatedBlockComment(sourceLocationAtStart()))
   }
 
+
+  /** Creates a [[SourceLocation]] of the inclusive `start` and exclusive `end` in the current source. */
+  private def mkSourceLocation(start: SourcePosition, end: SourcePosition)(implicit s: State): SourceLocation =
+    SourceLocation(isReal = true, s.src, start, end)
+
+  /** Returns the [[SourcePosition]] at [[State.start]]. */
+  private def sourcePositionAtStart()(implicit s: State): SourcePosition =
+    SourcePosition.mkFromZeroIndexed(s.start.line, s.start.column)
+
+  /** Returns a single-width [[SourceLocation]] starting at [[State.start]]. */
+  private def sourceLocationAtStart()(implicit s: State): SourceLocation =
+    SourceLocation.point(isReal = true, s.src, sourcePositionAtStart())
+
+  /** Returns the [[SourcePosition]] at the current cursor position (might be out of bounds). */
+  private def sourcePositionAtCurrent()(implicit s: State): SourcePosition =
+    SourcePosition.mkFromZeroIndexed(s.sc.getLine, s.sc.getColumn)
+
+  /** Returns a single-width [[SourceLocation]] starting at the current position. */
+  private def sourceLocationAtCurrent()(implicit s: State): SourceLocation =
+    SourceLocation.point(isReal = true, s.src, sourcePositionAtCurrent())
+
+  /** Returns a [[SourceLocation]] spanning the current consumed input since the last token, exclusive of the current position. */
+  private def sourceLocationFromStart()(implicit s: State): SourceLocation = {
```

Comment:
This location, spanning the whole of the current token, is what allows better locations in the follow up.